File: isabelle.eclipse.ui/src/isabelle/eclipse/ui/editors/TextHyperlink.java
package isabelle.eclipse.ui.editors;

import isabelle.eclipse.ui.IsabelleUIPlugin;

import java.net.URI;

import org.eclipse.jface.text.IRegion;
import org.eclipse.jface.text.hyperlink.IHyperlink;
import org.eclipse.ui.IEditorPart;
import org.eclipse.ui.IWorkbenchPage;
import org.eclipse.ui.PartInitException;


/**
 * A hyperlink to a text editor. Opens the given URI and selects target region in the opened editor.
 * 
 * @author Andrius Velykis
 */
public class TextHyperlink implements IHyperlink {

	private final IWorkbenchPage page;
	private final IRegion linkRegion;
	private final URI targetUri;
	private final String targetName;
	private final IRegion highlightRegion;
	private final IRegion selectRegion;
	
	public TextHyperlink(IWorkbenchPage page, IRegion linkRegion, URI targetUri, String targetName,
			IRegion selectRegion, IRegion highlightRegion) {
		
		this.page = page;
		this.linkRegion = linkRegion;
		this.targetUri = targetUri;
		this.targetName = targetName;
		this.highlightRegion = highlightRegion;
		this.selectRegion = selectRegion;
	}

	@Override
	public IRegion getHyperlinkRegion() {
		return linkRegion;
	}

	@Override
	public String getTypeLabel() {
		return "Text Link";
	}

	@Override
	public String getHyperlinkText() {
		return targetName;
	}

	@Override
	public void open() {
		
		IEditorPart editor;
		try {
			editor = EditorUtil.openEditor(page, targetUri);
		} catch (PartInitException e) {
			IsabelleUIPlugin.log(e.getLocalizedMessage(), e);
			return;
		}
		
		if (editor != null) {
			EditorUtil.revealInEditor(editor, selectRegion, highlightRegion);
		}
	}

}

File: isabelle.eclipse.ui/src/isabelle/eclipse/ui/editors/TextHyperlink.scala
package isabelle.eclipse.ui.editors

import java.net.URI

import org.eclipse.jface.text.IRegion
import org.eclipse.jface.text.hyperlink.IHyperlink
import org.eclipse.ui.{IWorkbenchPage, PartInitException}

import isabelle.eclipse.ui.internal.IsabelleUIPlugin.{error, log}


/**
 * A hyperlink to a text editor. 
 * 
 * Opens the given URI and selects target region in the opened editor.
 *
 * @param workbenchPage
 *            the page to open the editor in
 * @param linkRegion
 *            region in source editor representing the hyperlink location
 * @param targetName
 *            name of hyperlink
 * @param targetUri
 *            target URI to open in an editor
 * @param selectRegion
 *            a region within the target editor to be selected
 * @param highlightRegion
 *            a region within the target editor to be highlighted
 *            (normally the whole definition, while only the name is selected)
 *
 *
 * @author Andrius Velykis
 */
class TextHyperlink(
    workbenchPage: IWorkbenchPage,
    linkRegion: IRegion,
    targetName: Option[String],
    targetUri: URI,
    selectRegion: IRegion,
    highlightRegion: IRegion) extends IHyperlink {

  override def getHyperlinkRegion(): IRegion = linkRegion

  override def getTypeLabel = "Definition Link"

  override def getHyperlinkText(): String = targetName.orNull

  override def open() {

    try {

      val editor = EditorUtil.openEditor(workbenchPage, targetUri)
      EditorUtil.revealInEditor(editor, selectRegion, highlightRegion)

    } catch {
      case e: PartInitException => log(error(Some(e)))
    }
  }

}
